writer.h

```diff
@@ -46,4 +79,323 @@ class voxel_writer {
+		const H5std_string REGIONS_DATASET_NAME("regions");
+
+		hsize_t     regions_dimsf[1];
+		regions_dimsf[0] = continuous_count + constant_count + planar_count;
+
+		const int      REGION_RANK = 1;
+
+		H5::DataSpace regions_dataspace(REGION_RANK, regions_dimsf);
+		H5::PredType regions_datatype(H5::PredType::STD_REF_DSETREG);
+		regions_datatype.setOrder(H5T_ORDER_LE);
+		H5::DataSet regions_dataset = operation_group.createDataSet(REGIONS_DATASET_NAME, regions_datatype, regions_dataspace);
+
+		//  Regions dataset hyperslab preparation
+		hsize_t     region_offset[1];
+		region_offset[0] = -1;
+		hsize_t     region_slab_dimsf[1] = { 1 };
+		H5::DataSpace mspace3(REGION_RANK, region_slab_dimsf);
+
+		// Continuous chunks region hyperslab preparation
+		hsize_t     chunk_offset[4];
+		const int CHUNK_RANK = 4;
+		chunk_offset[0] = -1;
+		chunk_offset[1] = 0;
+		chunk_offset[2] = 0;
+		chunk_offset[3] = 0;
+		hsize_t     chunk_dimsf[4] = { 1, NX, NY, NZ };
+		H5::DataSpace chunk_space(CHUNK_RANK, chunk_dimsf);
+
+		// Planar chunks region hyperslab preparation 
+		hsize_t     planar_chunk_offset[1];
+		const int PLANAR_CHUNK_RANK = 1;
+		planar_chunk_offset[0] = 1;
+		hsize_t     planar_chunk_dimsf[1] = { 1 };
+		H5::DataSpace planar_chunk_space(PLANAR_CHUNK_RANK, planar_chunk_dimsf);
+
+		// Constant chunks region hyperslab preparation 
+		hsize_t     constant_chunk_offset[1];
+		const int CONSTANT_CHUNK_RANK = 1;
+		constant_chunk_offset[0] = 1;
+		hsize_t     constant_chunk_dimsf[1] = { 1 };
+		H5::DataSpace constant_chunk_space(CONSTANT_CHUNK_RANK, constant_chunk_dimsf);
+
+
+		BEGIN_LOOP(size_t(0), nchunks_x, 0U, nchunks_y, 0U, nchunks_z)
+			auto c = storage->get_chunk(ijk);
+		
+		
+
+		if (c == nullptr) {
+			
+			constant_offset[0]++;
+			constant_dataspace.selectHyperslab(H5S_SELECT_SET, constant_dims, constant_offset);
+			int constant_value[1] = { 0 };
+			dataset.write(constant_value, H5::PredType::NATIVE_INT, constant_space, constant_dataspace);
+
+
+			region_offset[0]++;
+			regions_dataspace.selectHyperslab(H5S_SELECT_SET, region_slab_dimsf, region_offset);
+
+			constant_chunk_offset[0]++;
+
+			hobj_ref_t inter[1];
+			file.reference(&inter[0], "/" + statement + "/constant_chunks", constant_dataspace, H5R_DATASET_REGION);
+			regions_dataset.write(inter, H5::PredType::STD_REF_DSETREG, mspace3, regions_dataspace);
+		}
+
+		else {
+
+			if (c->is_explicit()) {
+				
+				offset[0]++;
+			
+				size_t i0, j0, k0, i1, j1, k1;
+				i0 = 0;
+				j0 = 0;
+				k0 = 0;
+				c->extents().tie(i1, j1, k1);
+
+				BEGIN_LOOP_ZERO_2(make_vec<size_t>(i1, j1, k1))
+					offset[1] = ijk.get(0);
+					offset[2] = ijk.get(1);
+					offset[3] = ijk.get(2);
+					dataspace.selectHyperslab(H5S_SELECT_SET, slab_dimsf, offset);
+
+					if (c->value_bits() == 32) {
+
+						uint32_t  v;
+						std::vector<uint32_t> hslab;
+						c->Get(ijk, &v);
+						hslab.push_back(v);
+						dataset.write(hslab.data(), H5::PredType::NATIVE_INT, mspace2, dataspace);
```

Comment:
This doesn't seem correct, uint32_t is not the same as NATIVE_INT, I'm quite sure this will be completely wrong in 64bit systems. In case `c->value_bits() == 32` , the dataset type should be `H5::PredType::NATIVE_UINT32`